lib/ipam/ipam.go

```diff
@@ -922,10 +965,22 @@ func (c ipamClient) releaseIPsFromBlock(ctx context.Context, ips []net.IP, block
 			}
 		}
 
+		// List all handles, so we don't need to query them individually.
+		allHandles, err := c.blockReaderWriter.listHandles(ctx, "")
```

Comment:
This feels like an expensive operation to do per-block?  Could you pass in the map and then refresh it after a failure in the loop here?

I suppose you could also have a heuristic: if there's only on IP to free, query only that handle.